spec.html

```diff
@@ -32565,23 +32566,36 @@ <h1>Runtime Semantics: RegExpInitialize ( _obj_, _pattern_, _flags_ )</h1>
             1. If _flags_ is *undefined*, let _F_ be the empty String.
             1. Else, let _F_ be ? ToString(_flags_).
             1. If _F_ contains any code unit other than *"g"*, *"i"*, *"m"*, *"s"*, *"u"*, or *"y"* or if it contains the same code unit more than once, throw a *SyntaxError* exception.
-            1. If _F_ contains *"u"*, let _BMP_ be *false*; else let _BMP_ be *true*.
-            1. If _BMP_ is *true*, then
-              1. Let _pText_ be the sequence of code points resulting from interpreting each of the 16-bit elements of _P_ as a Unicode BMP code point. UTF-16 decoding is not applied to the elements.
-              1. Parse _pText_ using the grammars in <emu-xref href="#sec-patterns"></emu-xref>. The goal symbol for the parse is |Pattern[~U, ~N]|. If the result of parsing contains a |GroupName|, reparse with the goal symbol |Pattern[~U, +N]| and use this result instead. Throw a *SyntaxError* exception if _pText_ did not conform to the grammar, if any elements of _pText_ were not matched by the parse, or if any Early Error conditions exist.
-              1. Let _patternCharacters_ be a List whose elements are the code unit elements of _P_.
+            1. If _F_ contains *"u"*, let _u_ be *true*; else let _u_ be *false*.
+            1. If _u_ is *true*, then
+              1. Let _patternText_ be ! UTF16DecodeString(_P_).
+              1. Let _patternCharacters_ be a List whose elements are the code points of _patternText_.
             1. Else,
-              1. Let _pText_ be ! UTF16DecodeString(_P_).
-              1. Parse _pText_ using the grammars in <emu-xref href="#sec-patterns"></emu-xref>. The goal symbol for the parse is |Pattern[+U, +N]|. Throw a *SyntaxError* exception if _pText_ did not conform to the grammar, if any elements of _pText_ were not matched by the parse, or if any Early Error conditions exist.
-              1. Let _patternCharacters_ be a List whose elements are the code points of _pText_.
+              1. Let _patternText_ be the result of interpreting each of _P_'s 16-bit elements as a Unicode BMP code point. UTF-16 decoding is not applied to the elements.
+              1. Let _patternCharacters_ be a List whose elements are the code unit elements of _P_.
+            1. Let _parseResult_ be ParsePattern(_patternText_, _u_).
+            1. If _parseResult_ is a List of *SyntaxError* objects, throw a *SyntaxError* exception.
```

Comment:
what if it's a List of something else, or an empty List?

It seems like `ParsePattern` returns either a Parse Node, or not - should this be "is not a Parse Node"?

Reply:
It returns either a Parse Node or a List of SyntaxError objects. There's no other case.

We could follow this with an `Assert: _parseResult_ is a Parse Node`, if you think that's valuable.

Reply:
Fair, it just seems too vague to me to assert that it's a List *of syntax error objects* (without also reading the abstract operation), because of the possibilities that implies.

Reply:
@ljharb Would adding `Assert: _parseResult_ is a Parse Node` as the following line resolve your concern?

Reply:
I think I'd prefer changing the condition to "is not a Parse Node", as suggested by @ljharb. It's semantically equivalent, but easier on the reader: you don't need to look at the definition of ParsePattern to know that the only other possibility is that it's a Parse Node. (And we don't need to Assert that it is.)

Reply:
I kind of prefer writing it this way because it suggests that the error thrown could be one of the errors in the list, though I don't feel particularly strongly about it.

Reply:
In that case, should we say "throw one of the items in the list as an exception"? The impl has control of what's in the list anyways, so it wouldn't be normative, but would convey the intention you find important.

Reply:
There's a little bit of awkwardness there in that for that to be sensible the list has to be non-empty. Which it is guaranteed to be by the implementation of ParsePattern, but you have to look there to know that.

I'd be fine with that, I guess, but would still personally weakly lean towards the `If _parseResult_ is a List of *SyntaxError* objects, throw a *SyntaxError* exception` form + an assert.

But any of the suggestions here are adequate. I would mostly like us to just pick one and land it.

Reply:
I think it's valuable to follow up this step with an assertion. I don't care which way around the test is.

Reply:
I feel like any reasonable non-formalist reading of "a List of x objects" excludes the vacuous empty List case, but it doesn't hurt to qualify "a non-empty List of x objects", I guess.

And I agree with @bakkot on:
> I kind of prefer writing it this way because it suggests that the error thrown could be one of the errors in the list

So I'd prefer directly talking about errors than asserting it's not a parse node.

Reply:
I should add I'm happy with an _additional_ assert, but I like the condition for throwing checking for a list of errors over checking for not being a parse node.